[CodeGen] Update for scalable MemoryType in MMO

harviniriawan opened the pull request:
This is a continuation of #69152.

      Remove getSizeOrUnknown call when MachineMemOperand is created.
      For Scalable TypeSize, the MemoryType created becomes a
      scalable_vector.
    
      2 MMOs that have scalable memory access can then use the updated
      BasicAA that understands scalable LocationSize

davemgreen commented:
Hello. With Harvin having rotated back to a hardware team, I've pushed some extra changes to clean this up a little, remove the patch from #69152 and rebase.

RKSimon commented on llvm/include/llvm/CodeGen/MachineMemOperand.h:
```diff
@@ -233,13 +239,15 @@ class MachineMemOperand {
-    return MemoryType.isValid() ? MemoryType.getSizeInBits() : ~UINT64_C(0);
+  TypeSize getSizeInBits() const {
+    return MemoryType.isValid() ? MemoryType.getSizeInBits()
+                                : TypeSize::Fixed(~UINT64_C(0));
```
Is there any equivalent of MemoryLocation::UnknownSize we can use here?

davemgreen replied:
Hmm. Not that I can seen in TypeSize, I'm afraid. Some places appear to pass MemoryLocation::UnknownSize assuming they are the same size.

I could try and use MemoryLocation::UnknownSize in more places, or try and change it to an optional<> if that is better? That looks like it might be quite a bit of work though

nikic replied:
Maybe this should be using LocationSize rather than TypeSize, which can represent an unknown size? I'm not sure to what degree the different cases that LocationSize can represent are relevant at the MI level.

arsenm commented on llvm/lib/CodeGen/GlobalISel/LoadStoreOpt.cpp:
```diff
@@ -249,10 +246,20 @@ bool GISelAddressing::instMayAlias(const MachineInstr &MI,
+    return true;
+
+  const bool BothNotScalable =
+      !(MUC0.NumBytes.isScalable() || MUC1.NumBytes.isScalable());
```
!isScalable && !isScalable

davemgreen commented:
I've been having a look into making it a LocationSize. The size internally is stored in the size of a LLT, so it is only the input and output from the MMO that would handle LocationSize. (And afterPointer would not be supported, but that is probably fine).

I've had a go at making it use LocationSize. There are obviously a lot more changes, as it can no longer rely upon the implicit conversions from TypeSize->unsigned or from UnknownSize->unsigned. Hopefully I handled them all correctly, some places make the assumption that values cannot be Unknown, or handle it strangely it they do.

arsenm reviewed (commented):
The reason the MMO uses an LLT is we also need the number of vector elements to correctly legalize vector loads and stores. Doesn't switching that to LocationSize lose that information?

arsenm commented:
> Do you mean if we changed it internally too? Yeah that would lose the info I think. I'm not sure if it could come from the type of the load/store instead?
> 

It cannot in the case of extending vector load/truncating vector store. The current legalization for load/store is a total mess. I've had a WIP patch rewriting it for about 4 years which I really need to get back to

davemgreen commented:
I was thinking that would depend on whether the number of vector elements is assumed to always be the same between MMO and load type. i.e. the MMO could just store the size of the loaded memory, the load might extend each vector element. I'm not sure if there are somehow times when the types do not match?

arsenm commented:
> I was thinking that would depend on whether the number of vector elements is assumed to always be the same between MMO and load type. i.e. the MMO could just store the size of the loaded memory, the load might extend each vector element. I'm not sure if there are somehow times when the types do not match?

The cases get ambiguous with sub-byte element vectors

davemgreen commented:
I've given this a rebase.

davemgreen committed:
[SelectionDAG] Update for scalable MemoryType in MMO

Remove getSizeOrUnknown call when MachineMemOperand is created.  For Scalable
TypeSize, the MemoryType created becomes a scalable_vector.

2 MMOs that have scalable memory access can then use the updated BasicAA that
understands scalable LocationSize.

Original Patch by Harvin Iriawan

davemgreen commented:
Thanks. I've rebased to remove all the other changes from #84751, now leaving the Scalable vector parts here.

hiraditya commented:
Can this patch cause #88799?

hiraditya commented:
Reverting this patch does fix the crash.
